File: sdks/python/v_sentinel_sdk/clients/threat_intel.py
"""
V-Sentinel Threat Intelligence API Client

Provides async methods for threat intelligence operations.
"""

from typing import Optional, List, Dict, Any
import httpx

from ..models.threat_intel import (
    IOC,
    IOCList,
    IOCType,
    IOCConfidence,
    IOCCheckResult,
    ThreatActor,
    ThreatActorList,
    Campaign,
    CampaignList,
)
from ..exceptions import (
    ValidationError,
    ResourceNotFoundError,
    APIError,
)


class ThreatIntelClient:
    """
    Client for threat intelligence operations in V-Sentinel.
    
    Provides methods for IOC management, threat actor research,
    and campaign tracking.
    """
    
    def __init__(self, client: httpx.AsyncClient, base_url: str):
        """
        Initialize the threat intelligence client.
        
        Args:
            client: HTTP client instance
            base_url: Base API URL
        """
        self._client = client
        self._base_url = base_url.rstrip("/")
        self._endpoint = f"{self._base_url}/threat-intel"
    
    # ==================== IOC Operations ====================
    
    async def list_iocs(
        self,
        *,
        ioc_type: Optional[IOCType] = None,
        confidence: Optional[IOCConfidence] = None,
        threat_actor: Optional[str] = None,
        malware_family: Optional[str] = None,
        tags: Optional[List[str]] = None,
        time_range: Optional[str] = None,
        sort_by: Optional[str] = None,
        sort_order: Optional[str] = "desc",
        limit: Optional[int] = 50,
        offset: Optional[int] = 0,
    ) -> IOCList:
        """
        List IOCs with optional filtering.
        
        Args:
            ioc_type: Filter by IOC type
            confidence: Filter by confidence level
            threat_actor: Filter by threat actor name
            malware_family: Filter by malware family
            tags: Filter by tags
            time_range: Time range filter (e.g., '24h', '7d', '30d')
            sort_by: Field to sort by
            sort_order: Sort order ('asc' or 'desc')
            limit: Maximum number of results
            offset: Offset for pagination
            
        Returns:
            IOCList with IOCs and pagination info
        """
        params: Dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        
        if ioc_type:
            params["type"] = ioc_type.value
        if confidence:
            params["confidence"] = confidence.value
        if threat_actor:
            params["threat_actor"] = threat_actor
        if malware_family:
            params["malware_family"] = malware_family
        if tags:
            params["tags"] = ",".join(tags)
        if time_range:
            params["time_range"] = time_range
        if sort_by:
            params["sort_by"] = sort_by
        if sort_order:
            params["sort_order"] = sort_order
        
        response = await self._client.get(
            f"{self._endpoint}/iocs",
            params=params,
        )
        response.raise_for_status()
        
        data = response.json()
        return IOCList(**data)
    
    async def get_ioc(self, ioc_id: str) -> IOC:
        """
        Get a specific IOC by ID.
        
        Args:
            ioc_id: Unique IOC identifier
            
        Returns:
            IOC details
            
        Raises:
            ResourceNotFoundError: If IOC not found
        """
        response = await self._client.get(
            f"{self._endpoint}/iocs/{ioc_id}",
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"IOC not found: {ioc_id}",
                resource_type="ioc",
                resource_id=ioc_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return IOC(**data)
    
    async def check_ioc(
        self,
        value: str,
        *,
        ioc_type: Optional[IOCType] = None,
    ) -> IOCCheckResult:
        """
        Check if an indicator is malicious.
        
        This is the primary method for checking IOCs against
        the threat intelligence database.
        
        Args:
            value: IOC value to check
            ioc_type: Optional IOC type (auto-detected if not provided)
            
        Returns:
            IOCCheckResult with threat intelligence data
        """
        params: Dict[str, Any] = {"value": value}
        
        if ioc_type:
            params["type"] = ioc_type.value
        
        response = await self._client.get(
            f"{self._endpoint}/check",
            params=params,
        )
        response.raise_for_status()
        
        data = response.json()
        return IOCCheckResult(**data)
    
    async def check_iocs_bulk(
        self,
        values: List[str],
        *,
        ioc_type: Optional[IOCType] = None,
    ) -> List[IOCCheckResult]:
        """
        Check multiple IOCs in bulk.
        
        Args:
            values: List of IOC values to check
            ioc_type: Optional IOC type for all values
            
        Returns:
            List of IOCCheckResult for each value
        """
        payload: Dict[str, Any] = {"values": values}
        
        if ioc_type:
            payload["type"] = ioc_type.value
        
        response = await self._client.post(
            f"{self._endpoint}/check/bulk",
            json=payload,
        )
        response.raise_for_status()
        
        data = response.json()
        return [IOCCheckResult(**result) for result in data.get("results", [])]
    
    async def add_ioc(
        self,
        value: str,
        ioc_type: IOCType,
        *,
        confidence: IOCConfidence = IOCConfidence.MEDIUM,
        threat_actor: Optional[str] = None,
        campaign: Optional[str] = None,
        malware_family: Optional[str] = None,
        tags: Optional[List[str]] = None,
        description: Optional[str] = None,
        references: Optional[List[str]] = None,
        expiration_days: Optional[int] = None,
    ) -> IOC:
        """
        Add a new IOC to the database.
        
        Args:
            value: IOC value
            ioc_type: Type of IOC
            confidence: Confidence level (default: MEDIUM)
            threat_actor: Associated threat actor
            campaign: Associated campaign
            malware_family: Associated malware family
            tags: List of tags
            description: Description of the IOC
            references: List of reference URLs
            expiration_days: Days until IOC expires
            
        Returns:
            Created IOC
        """
        payload: Dict[str, Any] = {
            "value": value,
            "type": ioc_type.value,
            "confidence": confidence.value,
        }
        
        if threat_actor:
            payload["threat_actor"] = threat_actor
        if campaign:
            payload["campaign"] = campaign
        if malware_family:
            payload["malware_family"] = malware_family
        if tags:
            payload["tags"] = tags
        if description:
            payload["description"] = description
        if references:
            payload["references"] = references
        if expiration_days:
            payload["expiration_days"] = expiration_days
        
        response = await self._client.post(
            f"{self._endpoint}/iocs",
            json=payload,
        )
        response.raise_for_status()
        
        data = response.json()
        return IOC(**data)
    
    async def update_ioc(
        self,
        ioc_id: str,
        *,
        confidence: Optional[IOCConfidence] = None,
        threat_actor: Optional[str] = None,
        campaign: Optional[str] = None,
        malware_family: Optional[str] = None,
        tags: Optional[List[str]] = None,
        description: Optional[str] = None,
    ) -> IOC:
        """
        Update an existing IOC.
        
        Args:
            ioc_id: Unique IOC identifier
            confidence: New confidence level
            threat_actor: New threat actor
            campaign: New campaign
            malware_family: New malware family
            tags: New tags list
            description: New description
            
        Returns:
            Updated IOC
        """
        payload: Dict[str, Any] = {}
        
        if confidence is not None:
            payload["confidence"] = confidence.value
        if threat_actor is not None:
            payload["threat_actor"] = threat_actor
        if campaign is not None:
            payload["campaign"] = campaign
        if malware_family is not None:
            payload["malware_family"] = malware_family
        if tags is not None:
            payload["tags"] = tags
        if description is not None:
            payload["description"] = description
        
        response = await self._client.patch(
            f"{self._endpoint}/iocs/{ioc_id}",
            json=payload,
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"IOC not found: {ioc_id}",
                resource_type="ioc",
                resource_id=ioc_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return IOC(**data)
    
    async def delete_ioc(self, ioc_id: str) -> bool:
        """
        Delete an IOC.
        
        Args:
            ioc_id: Unique IOC identifier
            
        Returns:
            True if deleted successfully
        """
        response = await self._client.delete(
            f"{self._endpoint}/iocs/{ioc_id}",
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"IOC not found: {ioc_id}",
                resource_type="ioc",
                resource_id=ioc_id,
            )
        
        response.raise_for_status()
        return True
    
    async def search_iocs(
        self,
        query: str,
        *,
        limit: Optional[int] = 50,
    ) -> IOCList:
        """
        Search IOCs by query.
        
        Args:
            query: Search query
            limit: Maximum number of results
            
        Returns:
            IOCList with matching IOCs
        """
        response = await self._client.get(
            f"{self._endpoint}/iocs/search",
            params={"query": query, "limit": limit},
        )
        response.raise_for_status()
        
        data = response.json()
        return IOCList(**data)
    
    # ==================== Threat Actor Operations ====================
    
    async def list_threat_actors(
        self,
        *,
        country: Optional[str] = None,
        motivation: Optional[str] = None,
        tags: Optional[List[str]] = None,
        limit: Optional[int] = 50,
        offset: Optional[int] = 0,
    ) -> ThreatActorList:
        """
        List threat actors with optional filtering.
        
        Args:
            country: Filter by country of origin
            motivation: Filter by motivation type
            tags: Filter by tags
            limit: Maximum number of results
            offset: Offset for pagination
            
        Returns:
            ThreatActorList with actors and pagination info
        """
        params: Dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        
        if country:
            params["country"] = country
        if motivation:
            params["motivation"] = motivation
        if tags:
            params["tags"] = ",".join(tags)
        
        response = await self._client.get(
            f"{self._endpoint}/actors",
            params=params,
        )
        response.raise_for_status()
        
        data = response.json()
        return ThreatActorList(**data)
    
    async def get_threat_actor(self, actor_id: str) -> ThreatActor:
        """
        Get a specific threat actor by ID or name.
        
        Args:
            actor_id: Threat actor ID or name
            
        Returns:
            ThreatActor details
        """
        response = await self._client.get(
            f"{self._endpoint}/actors/{actor_id}",
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"Threat actor not found: {actor_id}",
                resource_type="threat_actor",
                resource_id=actor_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return ThreatActor(**data)
    
    async def get_threat_actor_iocs(
        self,
        actor_id: str,
        *,
        limit: Optional[int] = 100,
    ) -> IOCList:
        """
        Get IOCs associated with a threat actor.
        
        Args:
            actor_id: Threat actor ID or name
            limit: Maximum number of results
            
        Returns:
            IOCList with actor's IOCs
        """
        response = await self._client.get(
            f"{self._endpoint}/actors/{actor_id}/iocs",
            params={"limit": limit},
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"Threat actor not found: {actor_id}",
                resource_type="threat_actor",
                resource_id=actor_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return IOCList(**data)
    
    # ==================== Campaign Operations ====================
    
    async def list_campaigns(
        self,
        *,
        threat_actor: Optional[str] = None,
        status: Optional[str] = None,
        time_range: Optional[str] = None,
        limit: Optional[int] = 50,
        offset: Optional[int] = 0,
    ) -> CampaignList:
        """
        List campaigns with optional filtering.
        
        Args:
            threat_actor: Filter by threat actor
            status: Filter by campaign status
            time_range: Time range filter
            limit: Maximum number of results
            offset: Offset for pagination
            
        Returns:
            CampaignList with campaigns and pagination info
        """
        params: Dict[str, Any] = {
            "limit": limit,
            "offset": offset,
        }
        
        if threat_actor:
            params["threat_actor"] = threat_actor
        if status:
            params["status"] = status
        if time_range:
            params["time_range"] = time_range
        
        response = await self._client.get(
            f"{self._endpoint}/campaigns",
            params=params,
        )
        response.raise_for_status()
        
        data = response.json()
        return CampaignList(**data)
    
    async def get_campaign(self, campaign_id: str) -> Campaign:
        """
        Get a specific campaign by ID.
        
        Args:
            campaign_id: Unique campaign identifier
            
        Returns:
            Campaign details
        """
        response = await self._client.get(
            f"{self._endpoint}/campaigns/{campaign_id}",
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"Campaign not found: {campaign_id}",
                resource_type="campaign",
                resource_id=campaign_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return Campaign(**data)
    
    async def get_campaign_iocs(
        self,
        campaign_id: str,
        *,
        limit: Optional[int] = 100,
    ) -> IOCList:
        """
        Get IOCs associated with a campaign.
        
        Args:
            campaign_id: Unique campaign identifier
            limit: Maximum number of results
            
        Returns:
            IOCList with campaign's IOCs
        """
        response = await self._client.get(
            f"{self._endpoint}/campaigns/{campaign_id}/iocs",
            params={"limit": limit},
        )
        
        if response.status_code == 404:
            raise ResourceNotFoundError(
                f"Campaign not found: {campaign_id}",
                resource_type="campaign",
                resource_id=campaign_id,
            )
        
        response.raise_for_status()
        
        data = response.json()
        return IOCList(**data)
    
    # ==================== Export Operations ====================
    
    async def export_iocs(
        self,
        *,
        ioc_type: Optional[IOCType] = None,
        threat_actor: Optional[str] = None,
        malware_family: Optional[str] = None,
        tags: Optional[List[str]] = None,
        format: str = "json",
        include_expired: bool = False,
    ) -> str:
        """
        Export IOCs in specified format.
        
        Args:
            ioc_type: Filter by IOC type
            threat_actor: Filter by threat actor
            malware_family: Filter by malware family
            tags: Filter by tags
            format: Export format ('json', 'stix', 'csv', 'txt')
            include_expired: Include expired IOCs
            
        Returns:
            Exported data as string
        """
        params: Dict[str, Any] = {
            "format": format,
            "include_expired": include_expired,
        }
        
        if ioc_type:
            params["type"] = ioc_type.value
        if threat_actor:
            params["threat_actor"] = threat_actor
        if malware_family:
            params["malware_family"] = malware_family
        if tags:
            params["tags"] = ",".join(tags)
        
        response = await self._client.get(
            f"{self._endpoint}/export",
            params=params,
        )
        response.raise_for_status()
        
        return response.text
    
    async def export_for_siEM(
        self,
        siem_type: str,
        *,
        ioc_types: Optional[List[IOCType]] = None,
        threat_actor: Optional[str] = None,
    ) -> str:
        """
        Export IOCs formatted for specific SIEM.
        
        Args:
            siem_type: SIEM type ('splunk', 'elastic', 'sentinel', 'qradar')
            ioc_types: Filter by IOC types
            threat_actor: Filter by threat actor
            
        Returns:
            SIEM-formatted data
        """
        params: Dict[str, Any] = {"siem": siem_type}
        
        if ioc_types:
            params["types"] = ",".join(t.value for t in ioc_types)
        if threat_actor:
            params["threat_actor"] = threat_actor
        
        response = await self._client.get(
            f"{self._endpoint}/export/siem",
            params=params,
        )
        response.raise_for_status()
        
        return response.text

File: V-Sentinel-SDKs/typescript/src/services/threat-intel.ts
/**
 * Threat Intelligence API Service
 *
 * Handles threat intelligence API operations.
 */

import { AxiosInstance } from 'axios';
import { handleAPIError } from '../client';
import {
  IOC,
  IOCList,
  IOCType,
  IOCConfidence,
  IOCCheckResult,
  ThreatActor,
  ThreatActorList,
  Campaign,
  CampaignList,
} from '../models';

/** Options for listing IOCs */
export interface IOCListOptions {
  type?: IOCType;
  confidence?: IOCConfidence;
  threatActor?: string;
  malwareFamily?: string;
  tags?: string[];
  timeRange?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

/** Options for adding an IOC */
export interface AddIOCOptions {
  value: string;
  type: IOCType;
  confidence?: IOCConfidence;
  threatActor?: string;
  campaign?: string;
  malwareFamily?: string;
  tags?: string[];
  description?: string;
  references?: string[];
  expirationDays?: number;
}

/** Options for updating an IOC */
export interface UpdateIOCOptions {
  confidence?: IOCConfidence;
  threatActor?: string;
  campaign?: string;
  malwareFamily?: string;
  tags?: string[];
  description?: string;
}

/** Options for listing threat actors */
export interface ThreatActorListOptions {
  country?: string;
  motivation?: string;
  tags?: string[];
  limit?: number;
  offset?: number;
}

/** Options for listing campaigns */
export interface CampaignListOptions {
  threatActor?: string;
  status?: string;
  timeRange?: string;
  limit?: number;
  offset?: number;
}

/** Options for exporting IOCs */
export interface ExportOptions {
  type?: IOCType;
  threatActor?: string;
  malwareFamily?: string;
  tags?: string[];
  format?: 'json' | 'stix' | 'csv' | 'txt';
  includeExpired?: boolean;
}

/** Threat Intelligence API service */
export class ThreatIntelService {
  private readonly client: AxiosInstance;
  private readonly basePath = '/threat-intel';

  constructor(client: AxiosInstance) {
    this.client = client;
  }

  // ==================== IOC Operations ====================

  /**
   * List IOCs with optional filtering
   */
  async listIOCs(options?: IOCListOptions): Promise<IOCList> {
    try {
      const params: Record<string, string | number> = {};

      if (options?.type) params.type = options.type;
      if (options?.confidence) params.confidence = options.confidence;
      if (options?.threatActor) params.threat_actor = options.threatActor;
      if (options?.malwareFamily) params.malware_family = options.malwareFamily;
      if (options?.tags?.length) params.tags = options.tags.join(',');
      if (options?.timeRange) params.time_range = options.timeRange;
      if (options?.sortBy) params.sort_by = options.sortBy;
      if (options?.sortOrder) params.sort_order = options.sortOrder;
      if (options?.limit) params.limit = options.limit;
      if (options?.offset) params.offset = options.offset;

      if (!params.limit) params.limit = 50;

      const response = await this.client.get<IOCList>(`${this.basePath}/iocs`, { params });
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get a specific IOC by ID
   */
  async getIOC(iocId: string): Promise<IOC> {
    try {
      const response = await this.client.get<IOC>(
        `${this.basePath}/iocs/${iocId}`
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Check if an indicator is malicious
   */
  async checkIOC(value: string, type?: IOCType): Promise<IOCCheckResult> {
    try {
      const params: Record<string, string> = { value };
      if (type) params.type = type;

      const response = await this.client.get<IOCCheckResult>(
        `${this.basePath}/check`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Check multiple IOCs in bulk
   */
  async checkIOCsBulk(values: string[], type?: IOCType): Promise<IOCCheckResult[]> {
    try {
      const body: Record<string, unknown> = { values };
      if (type) body.type = type;

      const response = await this.client.post<{ results: IOCCheckResult[] }>(
        `${this.basePath}/check/bulk`,
        body
      );
      return response.data.results;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Add a new IOC to the database
   */
  async addIOC(options: AddIOCOptions): Promise<IOC> {
    try {
      const body: Record<string, unknown> = {
        value: options.value,
        type: options.type,
        confidence: options.confidence ?? IOCConfidence.Medium,
      };

      if (options.threatActor) body.threat_actor = options.threatActor;
      if (options.campaign) body.campaign = options.campaign;
      if (options.malwareFamily) body.malware_family = options.malwareFamily;
      if (options.tags?.length) body.tags = options.tags;
      if (options.description) body.description = options.description;
      if (options.references?.length) body.references = options.references;
      if (options.expirationDays) body.expiration_days = options.expirationDays;

      const response = await this.client.post<IOC>(`${this.basePath}/iocs`, body);
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Update an existing IOC
   */
  async updateIOC(iocId: string, options: UpdateIOCOptions): Promise<IOC> {
    try {
      const body: Record<string, unknown> = {};

      if (options.confidence) body.confidence = options.confidence;
      if (options.threatActor) body.threat_actor = options.threatActor;
      if (options.campaign) body.campaign = options.campaign;
      if (options.malwareFamily) body.malware_family = options.malwareFamily;
      if (options.tags) body.tags = options.tags;
      if (options.description) body.description = options.description;

      const response = await this.client.patch<IOC>(
        `${this.basePath}/iocs/${iocId}`,
        body
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Delete an IOC
   */
  async deleteIOC(iocId: string): Promise<void> {
    try {
      await this.client.delete(`${this.basePath}/iocs/${iocId}`);
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Search for IOCs matching a query
   */
  async searchIOCs(query: string, limit?: number): Promise<IOCList> {
    try {
      const params: Record<string, string | number> = { query };
      if (limit) params.limit = limit;
      else params.limit = 50;

      const response = await this.client.get<IOCList>(
        `${this.basePath}/iocs/search`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  // ==================== Threat Actor Operations ====================

  /**
   * List threat actors with optional filtering
   */
  async listThreatActors(options?: ThreatActorListOptions): Promise<ThreatActorList> {
    try {
      const params: Record<string, string | number> = {};

      if (options?.country) params.country = options.country;
      if (options?.motivation) params.motivation = options.motivation;
      if (options?.tags?.length) params.tags = options.tags.join(',');
      if (options?.limit) params.limit = options.limit;
      if (options?.offset) params.offset = options.offset;

      if (!params.limit) params.limit = 50;

      const response = await this.client.get<ThreatActorList>(
        `${this.basePath}/actors`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get a specific threat actor by ID or name
   */
  async getThreatActor(actorId: string): Promise<ThreatActor> {
    try {
      const response = await this.client.get<ThreatActor>(
        `${this.basePath}/actors/${actorId}`
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get IOCs associated with a threat actor
   */
  async getThreatActorIOCs(actorId: string, limit?: number): Promise<IOCList> {
    try {
      const params: Record<string, number> = {};
      if (limit) params.limit = limit;
      else params.limit = 100;

      const response = await this.client.get<IOCList>(
        `${this.basePath}/actors/${actorId}/iocs`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  // ==================== Campaign Operations ====================

  /**
   * List campaigns with optional filtering
   */
  async listCampaigns(options?: CampaignListOptions): Promise<CampaignList> {
    try {
      const params: Record<string, string | number> = {};

      if (options?.threatActor) params.threat_actor = options.threatActor;
      if (options?.status) params.status = options.status;
      if (options?.timeRange) params.time_range = options.timeRange;
      if (options?.limit) params.limit = options.limit;
      if (options?.offset) params.offset = options.offset;

      if (!params.limit) params.limit = 50;

      const response = await this.client.get<CampaignList>(
        `${this.basePath}/campaigns`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get a specific campaign by ID
   */
  async getCampaign(campaignId: string): Promise<Campaign> {
    try {
      const response = await this.client.get<Campaign>(
        `${this.basePath}/campaigns/${campaignId}`
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get IOCs associated with a campaign
   */
  async getCampaignIOCs(campaignId: string, limit?: number): Promise<IOCList> {
    try {
      const params: Record<string, number> = {};
      if (limit) params.limit = limit;
      else params.limit = 100;

      const response = await this.client.get<IOCList>(
        `${this.basePath}/campaigns/${campaignId}/iocs`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  // ==================== Export Operations ====================

  /**
   * Export IOCs in the specified format
   */
  async exportIOCs(options?: ExportOptions): Promise<string> {
    try {
      const params: Record<string, string | boolean> = {
        format: options?.format ?? 'json',
      };

      if (options?.type) params.type = options.type;
      if (options?.threatActor) params.threat_actor = options.threatActor;
      if (options?.malwareFamily) params.malware_family = options.malwareFamily;
      if (options?.tags?.length) params.tags = options.tags.join(',');
      if (options?.includeExpired) params.include_expired = options.includeExpired;

      const response = await this.client.get<string>(`${this.basePath}/export`, {
        params,
        responseType: 'text',
      });
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Export IOCs formatted for a specific SIEM
   */
  async exportForSIEM(
    siemType: 'splunk' | 'elastic' | 'sentinel' | 'qradar',
    iocTypes?: IOCType[],
    threatActor?: string
  ): Promise<string> {
    try {
      const params: Record<string, string> = { siem: siemType };

      if (iocTypes?.length) params.types = iocTypes.join(',');
      if (threatActor) params.threat_actor = threatActor;

      const response = await this.client.get<string>(`${this.basePath}/export/siem`, {
        params,
        responseType: 'text',
      });
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }
}